tests/wpt/metadata/xhr/formdata-foreach.html.ini

```diff
@@ -6,9 +6,6 @@
   [Entries iterator should return duplicate keys and non-deleted values]
     expected: FAIL
 
-  [Keys iterator should return duplicates]
-    expected: FAIL
-
   [Values iterator should return non-deleted values]
```

Comment:
About this test case, currently we only pass the `Keys` one because we failed to compare the `File` object.

After checking the object in detail, I found two points confused for me.

1. About `filename`

In spec, when [create an entry](https://xhr.spec.whatwg.org/#create-an-entry),

> If value is (now) a File object and filename is given, then set value to a new File object, representing the same bytes, whose name attribute value is filename.

It only says the filename should be followed by the `filename` passed to `append` or `set` function. However, both Gecko and Blink tries to get the _filename_ from `File` object if there's no passed filename. Should the spec be updated?

2. About `lastModified`

The `lastModified` between `File` object created by `new File` and the `File` created by the `create an entry` algorithm will be different.

Because we just used the `new` method from `File`,

https://github.com/servo/servo/blob/7c8b3eb94ad0082eae9d4262e19918dd3e7cb480/components/script/dom/formdata.rs#L188-L194

it will calculate the `lastModified` once we `new` it

https://github.com/servo/servo/blob/7c8b3eb94ad0082eae9d4262e19918dd3e7cb480/components/script/dom/file.rs#L37-L44

After doing some investigation, I found both Gecko and Blink has a method called `ToFile`.
- https://searchfox.org/mozilla-central/source/dom/base/FormData.cpp#28
- https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/html/forms/form_data.cc?l=359

With casting `Blob` to `File`, we should have same attributes for the object. I think we will do the same trick with `BlobImpl::File`?

If yes, I think maybe we need to update `FileBlob` to have a `modified` (or `lastModified`?) attribute so that we can check if it already has modified or not?

https://github.com/servo/servo/blob/7c8b3eb94ad0082eae9d4262e19918dd3e7cb480/components/script/dom/blob.rs#L24-L31

Reply:
Looks like Gecko at least stores the last modified time in BlobImpl effectively: https://searchfox.org/mozilla-central/rev/7ccc9d8b0bd9f879b6f24f15d8416493228697ac/dom/file/BaseBlobImpl.h#179

Reply:
How about storing the `BlobImpl::Memory` with a new struct ? so that we can reuse the `name` and `last_modified`.

```rust
#[derive(JSTraceable)]
pub struct MemoryBlob {
    name: Option<DOMString>,
    cache: Vec<u8>,
    last_modified: Option<i64>,
}

#[derive(JSTraceable)]
pub enum BlobImpl {
    ...
    Memory(MemoryBlob),
    ...
}

impl BlobImpl {
    /// Construct memory-backed BlobImpl
    #[allow(unrooted_must_root)]
    pub fn new_from_bytes(name: DOMString, bytes: Vec<u8>, last_modified: i64) -> BlobImpl {
        BlobImpl::Memory(MemoryBlob {
            name,
            cache: bytes,
            last_modified
        })
    }
}
```

Reply:
Filed at #22744